Update WooCommerce Analytics REST API Namespaces to lazy load for improved performance.

prettyboymp opened the pull request:
### Changes proposed in this Pull Request:

Improves WOOPLUG-2074 #45729.
Related: #60681 

This PR introduces a new ~`wc_rest_lazy_load_namespace()` function~ `RestApiUtil::lazy_load_namespace()` method that allows the controllers for a REST API namespace to be lazy loaded on demand.  It also, switches the ~`wc-admin` and~ `wc-analytics` namespaces to lazy load.  

Local testing shows that the loading of the WooCommerce Admin API can add over 100ms to the TTFB for API requests, all falling within `Automattic\WooCommerce\Admin\API\Init::rest_api_init()`

<img width="1243" height="420" alt="Screenshot 2025-09-02 at 11 12 34 AM" src="https://github.com/user-attachments/assets/d4199138-e796-4453-ae3a-6adaff40ee96" />

There were previous attempts, e.g. #47704, to improve this by only loading some API namespaces when it was needed by introducing `wc_rest_should_load_namespace()`.  However, because any code run during an API request can call `rest_do_request()`, the implementation had to be pretty limited and only skip loading when the current request was for a known API endpoint that didn't interact with another known namespace.  This limited the impact and still meant that the Admin API endpoints were still loading for most core REST requests.

This is a new attempt to deal with `rest_do_request()` by hooking into `rest_pre_dispatch` to late load an API namespace if the new requested route matches.



### Screenshots or screen recordings:

For comparison, I gathered some of the URLs that were currently forcing the WC Admin REST APIs to load.  Some of these actually used the API and some didn't.  I then created a performance test in K6 to hit each URL 20 times to compare the before and after.

Before (Trunk): 

```
checks.................................................................................................................................................................................: 100.00% ✓ 1240     ✗ 0
data_received..........................................................................................................................................................................: 141 MB  240 kB/s
data_sent..............................................................................................................................................................................: 796 kB  1.4 kB/s
group_duration.........................................................................................................................................................................: avg=415.86ms   min=119.93ms   med=238.10ms   max=3952.41ms  p(90)=1118.70ms  p(95)=1272.12ms
http_req_blocked.......................................................................................................................................................................: avg=0.01ms     min=0.00ms     med=0.00ms     max=1.59ms     p(90)=0.01ms     p(95)=0.01ms
http_req_connecting....................................................................................................................................................................: avg=0.00ms     min=0.00ms     med=0.00ms     max=0.22ms     p(90)=0.00ms     p(95)=0.00ms
http_req_duration......................................................................................................................................................................: avg=387.63ms   min=0.65ms     med=233.14ms   max=3740.63ms  p(90)=1099.51ms  p(95)=1251.77ms
 { expected_response:true }...........................................................................................................................................................: avg=387.63ms   min=0.65ms     med=233.14ms   max=3740.63ms  p(90)=1099.51ms  p(95)=1251.77ms
✓ { group:::/wp-admin/admin.php?page=wc-admin&path=/analytics/overview }...............................................................................................................: avg=422.44ms   min=335.65ms   med=357.89ms   max=1367.50ms  p(90)=435.27ms   p(95)=533.69ms
✓ { group:::/wp-admin/admin.php?page=wc-orders }.......................................................................................................................................: avg=1029.80ms  min=955.45ms   med=1023.73ms  max=1116.56ms  p(90)=1083.32ms  p(95)=1093.45ms
✓ { group:::/wp-admin/admin.php?page=wc-settings&tab=checkout&from=PAYMENTS_MENU_ITEM }................................................................................................: avg=366.61ms   min=335.57ms   med=358.41ms   max=480.24ms   p(90)=398.14ms   p(95)=435.48ms
✓ { group:::/wp-admin/post.php?post=1103554&action=edit }..............................................................................................................................: avg=1256.55ms  min=1094.81ms  med=1178.75ms  max=2285.61ms  p(90)=1424.80ms  p(95)=1711.47ms
✓ { group:::/wp-admin/post.php?post=1270814&action=edit }..............................................................................................................................: avg=1221.05ms  min=1118.65ms  med=1214.58ms  max=1393.65ms  p(90)=1303.36ms  p(95)=1340.73ms
✓ { group:::/wp-json/jetpack/v4/connection?_locale=user }..............................................................................................................................: avg=234.14ms   min=218.76ms   med=232.00ms   max=271.71ms   p(90)=249.22ms   p(95)=251.10ms
✓ { group:::/wp-json/wc-admin/marketing/misc-recommendations?_locale=user }............................................................................................................: avg=135.19ms   min=119.73ms   med=133.43ms   max=186.18ms   p(90)=145.10ms   p(95)=150.89ms
✓ { group:::/wp-json/wc-admin/onboarding/tasks?_locale=user }..........................................................................................................................: avg=222.32ms   min=206.92ms   med=215.82ms   max=314.54ms   p(90)=234.63ms   p(95)=285.10ms
✓ { group:::/wp-json/wc-admin/options?options=woocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp&_locale=user }...........................................: avg=139.66ms   min=120.16ms   med=133.65ms   max=270.25ms   p(90)=153.95ms   p(95)=165.66ms
✓ { group:::/wp-json/wc-admin/options?options=woocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp%2Cwoocommerce_allow_tracking&_locale=user }..............: avg=139.26ms   min=120.89ms   med=131.54ms   max=277.63ms   p(90)=162.58ms   p(95)=167.42ms
✓ { group:::/wp-json/wc-admin/options?options=woocommerce_ces_tracks_queue&_locale=user }..............................................................................................: avg=140.17ms   min=122.49ms   med=129.17ms   max=332.13ms   p(90)=157.30ms   p(95)=174.39ms
✓ { group:::/wp-json/wc-admin/options?options=woocommerce_remote_variant_assignment%2Cwoocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp&_locale=user }...: avg=137.61ms   min=122.11ms   med=132.18ms   max=193.47ms   p(90)=169.04ms   p(95)=179.52ms
✓ { group:::/wp-json/wc-admin/plugins/active?_locale=user }............................................................................................................................: avg=136.83ms   min=120.81ms   med=132.90ms   max=183.66ms   p(90)=161.13ms   p(95)=167.59ms
✓ { group:::/wp-json/wc-admin/plugins/installed?_locale=user }.........................................................................................................................: avg=156.04ms   min=146.96ms   med=152.00ms   max=202.68ms   p(90)=172.94ms   p(95)=187.58ms
✓ { group:::/wp-json/wc-analytics/admin/notes?page=1&per_page=25&type=error%2Cupdate&status=unactioned&_locale=user }..................................................................: avg=193.41ms   min=168.94ms   med=184.15ms   max=347.98ms   p(90)=211.20ms   p(95)=243.56ms
✓ { group:::/wp-json/wc-analytics/leaderboards?after=2025-09-01T00%3A00%3A00&before=2025-09-02T23%3A59%3A59&per_page=5&persisted_query=%7B%7D&_locale=user }...........................: avg=187.34ms   min=171.13ms   med=183.23ms   max=221.91ms   p(90)=202.80ms   p(95)=210.72ms
✓ { group:::/wp-json/wp/v2/block-patterns/patterns?_locale=user }......................................................................................................................: avg=542.70ms   min=515.26ms   med=538.56ms   max=596.30ms   p(90)=581.26ms   p(95)=593.52ms
✓ { group:::/wp-json/wp/v2/blocks?context=edit&per_page=100&page=1&_locale=user }......................................................................................................: avg=249.76ms   min=224.70ms   med=238.78ms   max=389.00ms   p(90)=271.37ms   p(95)=292.35ms
✓ { group:::/wp-json/wp/v2/posts?_locale=user }........................................................................................................................................: avg=325.48ms   min=292.21ms   med=321.20ms   max=370.21ms   p(90)=364.30ms   p(95)=367.36ms
✓ { group:::/wp-json/wp/v2/settings?_locale=user }.....................................................................................................................................: avg=247.45ms   min=225.18ms   med=241.50ms   max=296.39ms   p(90)=273.11ms   p(95)=279.91ms
✓ { group:::/wp-json/wp/v2/taxonomies?context=edit&per_page=100&_locale=user }.........................................................................................................: avg=243.49ms   min=225.23ms   med=239.48ms   max=304.64ms   p(90)=256.94ms   p(95)=284.71ms
✓ { group:::/wp-json/wp/v2/templates?context=edit&per_page=100&_locale=user }..........................................................................................................: avg=247.88ms   min=221.21ms   med=238.46ms   max=363.74ms   p(90)=285.36ms   p(95)=307.08ms
✓ { group:::/wp-json/wp/v2/templates/lookup?slug=front-page&_locale=user }.............................................................................................................: avg=246.77ms   min=226.93ms   med=242.36ms   max=289.89ms   p(90)=264.76ms   p(95)=274.39ms
✓ { group:::/wp-json/wp/v2/users?context=view&who=authors&per_page=100&_fields=id%2Cname&_locale=user }................................................................................: avg=268.44ms   min=235.56ms   med=263.54ms   max=394.43ms   p(90)=292.14ms   p(95)=319.44ms
✓ { group:::/wp-json/wp/v2/users/1?context=view&_fields=id%2Cname&_locale=user }.......................................................................................................: avg=256.71ms   min=232.67ms   med=252.98ms   max=337.01ms   p(90)=267.67ms   p(95)=304.94ms
✓ { group:::/wp-json/wp/v2/wp_pattern_category?context=view&per_page=100&_fields=id%2Cname%2Cdescription%2Cslug&_locale=user }.........................................................: avg=250.23ms   min=225.45ms   med=239.87ms   max=376.33ms   p(90)=278.97ms   p(95)=300.09ms
http_req_failed........................................................................................................................................................................: 0.00%   ✓ 0        ✗ 1200
http_req_receiving.....................................................................................................................................................................: avg=7.59ms     min=0.02ms     med=0.14ms     max=1028.01ms  p(90)=5.51ms     p(95)=6.58ms
http_req_sending.......................................................................................................................................................................: avg=0.07ms     min=0.01ms     med=0.02ms     max=8.32ms     p(90)=0.05ms     p(95)=0.09ms
http_req_tls_handshaking...............................................................................................................................................................: avg=0.00ms     min=0.00ms     med=0.00ms     max=0.00ms     p(90)=0.00ms     p(95)=0.00ms
http_req_waiting.......................................................................................................................................................................: avg=379.96ms   min=0.59ms     med=231.70ms   max=2722.00ms  p(90)=1089.44ms  p(95)=1237.64ms
http_reqs..............................................................................................................................................................................: 1200    2.04096/s
iteration_duration.....................................................................................................................................................................: avg=14698.95ms min=11370.88ms med=14807.07ms max=18718.62ms p(90)=16414.15ms p(95)=16594.54ms
iterations.............................................................................................................................................................................: 40      0.068032/s
vus....................................................................................................................................................................................: 1       min=1      max=1
vus_max................................................................................................................................................................................: 1       min=1      max=1
```

After (Lazy Loaded Admin/Analytics API) :

```
checks.................................................................................................................................................................................: 100.00% ✓ 1240     ✗ 0
data_received..........................................................................................................................................................................: 141 MB  255 kB/s
data_sent..............................................................................................................................................................................: 796 kB  1.4 kB/s
group_duration.........................................................................................................................................................................: avg=386.99ms   min=119.13ms   med=182.12ms   max=3285.92ms  p(90)=1108.06ms  p(95)=1289.02ms
http_req_blocked.......................................................................................................................................................................: avg=0.01ms     min=0.00ms     med=0.01ms     max=3.75ms     p(90)=0.01ms     p(95)=0.01ms
http_req_connecting....................................................................................................................................................................: avg=0.00ms     min=0.00ms     med=0.00ms     max=1.98ms     p(90)=0.00ms     p(95)=0.00ms
http_req_duration......................................................................................................................................................................: avg=360.72ms   min=0.61ms     med=181.59ms   max=3100.95ms  p(90)=1088.95ms  p(95)=1250.81ms
 { expected_response:true }...........................................................................................................................................................: avg=360.72ms   min=0.61ms     med=181.59ms   max=3100.95ms  p(90)=1088.95ms  p(95)=1250.81ms
✓ { group:::/wp-admin/admin.php?page=wc-admin&path=/analytics/overview }...............................................................................................................: avg=437.90ms   min=356.94ms   med=377.82ms   max=1387.28ms  p(90)=446.36ms   p(95)=582.04ms
✓ { group:::/wp-admin/admin.php?page=wc-orders }.......................................................................................................................................: avg=1068.89ms  min=954.18ms   med=1063.15ms  max=1250.72ms  p(90)=1131.58ms  p(95)=1192.38ms
✓ { group:::/wp-admin/admin.php?page=wc-settings&tab=checkout&from=PAYMENTS_MENU_ITEM }................................................................................................: avg=461.06ms   min=352.72ms   med=378.78ms   max=1436.22ms  p(90)=444.25ms   p(95)=1383.68ms
✓ { group:::/wp-admin/post.php?post=1103554&action=edit }..............................................................................................................................: avg=1188.60ms  min=1037.85ms  med=1149.12ms  max=1493.83ms  p(90)=1441.18ms  p(95)=1489.41ms
✓ { group:::/wp-admin/post.php?post=1270814&action=edit }..............................................................................................................................: avg=1185.74ms  min=1087.12ms  med=1164.78ms  max=1449.87ms  p(90)=1255.97ms  p(95)=1302.61ms
✓ { group:::/wp-json/jetpack/v4/connection?_locale=user }..............................................................................................................................: avg=177.37ms   min=166.07ms   med=172.41ms   max=210.83ms   p(90)=196.69ms   p(95)=201.18ms
✓ { group:::/wp-json/wc-admin/marketing/misc-recommendations?_locale=user }............................................................................................................: avg=128.53ms   min=119.22ms   med=126.36ms   max=151.63ms   p(90)=137.16ms   p(95)=139.93ms
✓ { group:::/wp-json/wc-admin/onboarding/tasks?_locale=user }..........................................................................................................................: avg=221.94ms   min=201.63ms   med=208.86ms   max=348.72ms   p(90)=255.64ms   p(95)=284.56ms
✓ { group:::/wp-json/wc-admin/options?options=woocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp&_locale=user }...........................................: avg=134.20ms   min=124.10ms   med=131.04ms   max=174.78ms   p(90)=144.06ms   p(95)=153.24ms
✓ { group:::/wp-json/wc-admin/options?options=woocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp%2Cwoocommerce_allow_tracking&_locale=user }..............: avg=138.06ms   min=120.79ms   med=128.99ms   max=283.15ms   p(90)=147.25ms   p(95)=168.76ms
✓ { group:::/wp-json/wc-admin/options?options=woocommerce_ces_tracks_queue&_locale=user }..............................................................................................: avg=127.83ms   min=118.90ms   med=125.80ms   max=157.32ms   p(90)=134.88ms   p(95)=141.69ms
✓ { group:::/wp-json/wc-admin/options?options=woocommerce_remote_variant_assignment%2Cwoocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp&_locale=user }...: avg=129.54ms   min=120.17ms   med=126.00ms   max=168.15ms   p(90)=137.52ms   p(95)=155.87ms
✓ { group:::/wp-json/wc-admin/plugins/active?_locale=user }............................................................................................................................: avg=134.09ms   min=121.61ms   med=129.59ms   max=180.68ms   p(90)=149.73ms   p(95)=157.97ms
✓ { group:::/wp-json/wc-admin/plugins/installed?_locale=user }.........................................................................................................................: avg=160.06ms   min=142.55ms   med=151.26ms   max=201.23ms   p(90)=191.71ms   p(95)=194.62ms
✓ { group:::/wp-json/wc-analytics/admin/notes?page=1&per_page=25&type=error%2Cupdate&status=unactioned&_locale=user }..................................................................: avg=168.42ms   min=152.16ms   med=160.99ms   max=302.69ms   p(90)=182.02ms   p(95)=204.47ms
✓ { group:::/wp-json/wc-analytics/leaderboards?after=2025-09-01T00%3A00%3A00&before=2025-09-02T23%3A59%3A59&per_page=5&persisted_query=%7B%7D&_locale=user }...........................: avg=170.25ms   min=153.30ms   med=164.49ms   max=329.63ms   p(90)=177.77ms   p(95)=187.28ms
✓ { group:::/wp-json/wp/v2/block-patterns/patterns?_locale=user }......................................................................................................................: avg=488.04ms   min=454.59ms   med=473.50ms   max=676.13ms   p(90)=526.00ms   p(95)=534.49ms
✓ { group:::/wp-json/wp/v2/blocks?context=edit&per_page=100&page=1&_locale=user }......................................................................................................: avg=184.99ms   min=167.89ms   med=176.73ms   max=316.31ms   p(90)=197.79ms   p(95)=221.50ms
✓ { group:::/wp-json/wp/v2/posts?_locale=user }........................................................................................................................................: avg=253.68ms   min=231.55ms   med=247.49ms   max=353.80ms   p(90)=273.77ms   p(95)=287.67ms
✓ { group:::/wp-json/wp/v2/settings?_locale=user }.....................................................................................................................................: avg=200.24ms   min=165.67ms   med=182.53ms   max=343.08ms   p(90)=246.04ms   p(95)=332.46ms
✓ { group:::/wp-json/wp/v2/taxonomies?context=edit&per_page=100&_locale=user }.........................................................................................................: avg=179.53ms   min=165.65ms   med=175.26ms   max=213.34ms   p(90)=194.32ms   p(95)=201.41ms
✓ { group:::/wp-json/wp/v2/templates?context=edit&per_page=100&_locale=user }..........................................................................................................: avg=202.70ms   min=167.67ms   med=176.35ms   max=893.91ms   p(90)=203.12ms   p(95)=224.09ms
✓ { group:::/wp-json/wp/v2/templates/lookup?slug=front-page&_locale=user }.............................................................................................................: avg=191.12ms   min=172.68ms   med=182.88ms   max=305.78ms   p(90)=209.88ms   p(95)=239.81ms
✓ { group:::/wp-json/wp/v2/users?context=view&who=authors&per_page=100&_fields=id%2Cname&_locale=user }................................................................................: avg=197.41ms   min=180.44ms   med=192.03ms   max=228.71ms   p(90)=217.28ms   p(95)=219.99ms
✓ { group:::/wp-json/wp/v2/users/1?context=view&_fields=id%2Cname&_locale=user }.......................................................................................................: avg=194.13ms   min=181.21ms   med=193.33ms   max=223.88ms   p(90)=204.05ms   p(95)=209.13ms
✓ { group:::/wp-json/wp/v2/wp_pattern_category?context=view&per_page=100&_fields=id%2Cname%2Cdescription%2Cslug&_locale=user }.........................................................: avg=176.73ms   min=165.37ms   med=175.30ms   max=201.80ms   p(90)=183.97ms   p(95)=191.14ms
http_req_failed........................................................................................................................................................................: 0.00%   ✓ 0        ✗ 1200
http_req_receiving.....................................................................................................................................................................: avg=7.61ms     min=0.03ms     med=0.15ms     max=1019.50ms  p(90)=5.34ms     p(95)=6.49ms
http_req_sending.......................................................................................................................................................................: avg=0.07ms     min=0.01ms     med=0.03ms     max=6.32ms     p(90)=0.04ms     p(95)=0.10ms
http_req_tls_handshaking...............................................................................................................................................................: avg=0.00ms     min=0.00ms     med=0.00ms     max=0.00ms     p(90)=0.00ms     p(95)=0.00ms
http_req_waiting.......................................................................................................................................................................: avg=353.04ms   min=0.57ms     med=181.08ms   max=2509.00ms  p(90)=1078.21ms  p(95)=1227.55ms
http_reqs..............................................................................................................................................................................: 1200    2.163686/s
iteration_duration.....................................................................................................................................................................: avg=13864.90ms min=11330.76ms med=13989.11ms max=17414.93ms p(90)=15710.72ms p(95)=16343.21ms
iterations.............................................................................................................................................................................: 40      0.072123/s
vus....................................................................................................................................................................................: 1       min=1      max=1
vus_max................................................................................................................................................................................: 1       min=1      max=1
```

Analysis:
Most requests that didn't need the `wc-admin` or `wc-analytics` namespaces dropped between 30ms - 60ms (p90).

Of note, there is an increase in time of around 30 - 50ms for some other pages.  This is indirectly caused by the lazy loading of the WC Admin API's.  It doesn't happen because the APIs are lazy loaded, but because the Admin API's now load after other rest routes have been registered.  This causes the rest server to have to iterate through more routes before routing the multiple internal requests that the wc-analytics API makes when registering its routes.  Changing the registration in `Automattic\WooCommerce\Admin\API` to register the routes after other namespaces cause similar slow downs.  This will improve as we:

1. Move other APIs to be lazy loaded.
2. Address some of the specific pages that really don't need to be preloading some of the Amin API data.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

The biggest thing is smoke testing using different parts of the sites that load the API, either through direct requests or through preloading/`rest_do_request()`.  Areas I've tested include:

1. Page/Post editing with blocks, specifically those that pull in content like posts, categories, products, etc.
2. Site Editor
3. WooCommerce Analytics within the admin.

NOTE:  The following extensions currently use the `woocommerce_admin_rest_controllers` filter and ~should be~ have been tested.

- [x] AutomateWoo
- [x] AutomateWoo Referrals
- [x] WooCommerce Composite Products
- [x] WooCommerce Gift Cards
- [x] WooCommerce Product Bundles
- [x] WooCommerce Product Recommendations

-- Checking the code for each of these, they all append controllers to the filtered array, regardless of the context.  All of these are specifically analytics based controllers.  My thoughts are to deprecate the current filter and create two new filters while keeping the check to only run the filter once.

WooCommerce Payments should also test against these changes as there were previous issues with #47704:
* https://github.com/woocommerce/woocommerce/issues/50891
* https://github.com/Automattic/woocommerce-payments/pull/10505

<!-- End testing instructions -->

prettyboymp committed:
Update WooCommerce Admin REST API Namespaces to load on demand only.

prettyboymp committed:
Fixing callback argument order

prettyboymp committed:
Avoid skipping previously added controllers because it breaks tests that reset the server but not the Admin API service

prettyboymp committed:
Adjust tests
Add back loading of onboarding endpoint
Handle discovery at api root

prettyboymp committed:
Rename function to wc_rest_lazy_load_namespace to better identify what it does

prettyboymp committed:
Deal with jetpack mocking for performance indicator tests

prettyboymp committed:
Add tests and improve documentation

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

prettyboymp committed:
fix merge conflict resolution issue

prettyboymp committed:
lint fixes

prettyboymp committed:
more lint issues

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

prettyboymp committed:
Remove pending todo for handling woocommerce_admin_rest_controllers filter

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

prettyboymp committed:
lint fixes

prettyboymp commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,117 +54,148 @@ public function __construct() {
+			 * @since 3.5.0
+			 */
+			$controllers = apply_filters( 'woocommerce_admin_rest_controllers', $controllers );
 		}
```
Previously, this would only run once because both the wc-analytics and wc-admin namespaces would load under the same $contollers array.  Every usage of this filter I came across just added new controllers to the array, regardless of context.  So I chose to check if the filter already ran to avoid controllers being loaded twice.

prettyboymp commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,117 +54,148 @@
-		 * @since 3.5.0
-		 * @param array $controllers List of rest API controllers.
-		 */
-		$controllers = apply_filters( 'woocommerce_admin_rest_controllers', $controllers );
-
 		foreach ( $controllers as $controller ) {
 			$this->$controller = new $controller();
 			$this->$controller->register_routes();
 		}
 	}
```
I did this initially instead of forcing the filter to only run once, but it broke a bunch of tests because there isn't a great way to reset this class instance between tests.

prettyboymp committed:
Update plugins/woocommerce/src/Admin/API/Init.php

Remove duplicate ProductVariations controller

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

albarin commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,117 +54,147 @@ public function __construct() {
+	 */
+	public function rest_api_init_wc_analytics() {
+		// Controllers in wc-analytics namespace, but loaded irrespective of analytics feature value.
+		$controllers = array(
```
@prettyboymp These controllers won't be loaded if the `analytics` feature is disabled, because the loop with the `register_routes` happens only inside the `if ( Features::is_enabled( 'analytics' ))`.

To reproduce:
1. Go to `wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`.
2. Uncheck `Enable WooCommerce Analytics` and save
3. Go to `wp-admin/edit.php?post_type=shop_order`, open the console, and see the following errors.
<img width="1511" height="97" alt="Image" src="https://github.com/user-attachments/assets/11607d56-d1a2-4fd0-a412-7694cc1302d5" />
4. Enable analytics again, refresh the orders page, and see that they no longer appear.

prettyboymp replied:
This logic is fixed.  Thank you for catching.

albarin reviewed (changes requested):
Apart from the comment I shared before about analytics being disabled, everything else looks and tests good:

✅ Smoke tested pages and posts with blocks like single product, category list, etc.
✅ Smoke tested the site editor, in particular Woo templates (shop, cart, checkout, product, etc.).
✅ Smoke WooCommerce Analytics within the admin.

prettyboymp commented:
One thing I am seeing is that the way the REST_API merges/overwrites $route_args for later registration is going to be a problem if extensions expect to run after a current namespace is registered and suddenly aren't.  I wouldn't think that was extremely likely with most of the routes, but it looks like at least one built in test expects that:  https://github.com/woocommerce/woocommerce/blob/fcbfecb5f516db46dd596b45aa16010a22eb6568/plugins/woocommerce/tests/php/src/Blocks/StoreApi/Routes/Checkout.php#L75

Off hand I think we should still move forward with lazy loading the analytics and wc-admin namespaces, even if we split the change across multiple versions.  However, we probably need to consider feature flagging the lazy loading of other endpoints.  

Another option for improvement to consider: Add a filter at the end of some of these callbacks e.g. `wc-analytics_rest_namespace_routes_registered`, to allow other extensions to explicitly register and extend routes after core. This won't immediately fix conflicting extensions, but would give a better canonical way of extending/overriding WooCommerce core routes that is more future proof.  It would continue to work whether WordPress core adopted the ability to lazy load routes or we abandoned the idea altogether.

prettyboymp committed:
fix logic related to wc-analytics namespace loading when the analytics feature is disabled

prettyboymp committed:
Add action hooks immediately after each namespace is loaded so extensions can more easily target specific namespaces as we possibly change loading conditions

prettyboymp committed:
Only lazy load wc-analyics for now.

prettyboymp committed:
update testing comments around testing route registration to reflect that wc-admin may not necessarily be lazy loaded yet.

prettyboymp committed:
lint fixes

prettyboymp committed:
Correct when 'woocommerce_' . $namespace . '_rest_namespace_routes_registered' action is called within the REST API Server

prettyboymp committed:
fix duplicate request in test

prettyboymp committed:
more lint fixes

prettyboymp committed:
Remove test_profile_extensibility change as it looks to have been corrected upstream

prettyboymp committed:
restore wc_rest_should_load_namespace checkout around wc-admin api loading

prettyboymp commented:
I reduced down the changes in this PR to only add the new lazy loading function and lazy load the `wc-analytics` namespace.  

Here's the comprehensive comparison of p90 response times across **Trunk**, this patch in its current state with the Analytics Feature off, and this patch in its current state with the Analytics Feature enabled:

| URL | Trunk (ms) | Lazy Load WC Analytics (ms) | Lazy Load % Change | Feature Enabled (ms) | Feature % Change |
|-----|------------|----------------------------|-------------------|---------------------|------------------|
| /wp-admin/admin.php?page=wc-admin&path=/analytics/overview | 348.28 | **319.57** | **-8.24%** | 538.12 | +54.51% |
| /wp-admin/admin.php?page=wc-orders | **912.04** | 913.81 | +0.19% | 978.04 | +7.24% |
| /wp-admin/admin.php?page=wc-settings&tab=checkout&from=PAYMENTS_MENU_ITEM | 382.87 | **338.59** | **-11.57%** | 450.92 | +17.77% |
| /wp-admin/post.php?post=1103554&action=edit | **1251.71** | 1393.11 | +11.30% | 1297.27 | +3.64% |
| /wp-admin/post.php?post=1270814&action=edit | **1273.67** | 1308.74 | +2.75% | 1381.66 | +8.48% |
| /wp-json/jetpack/v4/connection?_locale=user | 209.41 | **204.25** | **-2.46%** | 211.47 | +0.98% |
| /wp-json/wc-admin/marketing/misc-recommendations?_locale=user | 147.81 | **147.70** | **-0.07%** | 157.68 | +6.68% |
| /wp-json/wc-admin/onboarding/tasks?_locale=user | **248.56** | 248.65 | +0.04% | 266.94 | +7.39% |
| /wp-json/wc-admin/options?options=woocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp&_locale=user | **149.98** | 158.88 | +5.93% | 157.21 | +4.82% |
| /wp-json/wc-admin/options?options=woocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp%2Cwoocommerce_allow_tracking&_locale=user | 153.99 | 152.81 | **-0.77%** | **145.47** | **-5.53%** |
| /wp-json/wc-admin/options?options=woocommerce_ces_tracks_queue&_locale=user | **146.42** | 151.66 | +3.58% | 148.98 | +1.75% |
| /wp-json/wc-admin/options?options=woocommerce_remote_variant_assignment%2Cwoocommerce_admin_transient_notices_queue%2Cwoocommerce_admin_install_timestamp&_locale=user | **145.17** | 148.07 | +2.00% | 152.67 | +5.17% |
| /wp-json/wc-admin/plugins/active?_locale=user | 155.47 | **150.68** | **-3.08%** | 159.42 | +2.54% |
| /wp-json/wc-admin/plugins/installed?_locale=user | **171.25** | 179.58 | +4.86% | 215.06 | +25.58% |
| /wp-json/wc-analytics/admin/notes?page=1&per_page=25&type=error%2Cupdate&status=unactioned&_locale=user | **160.74** | 163.71 | +1.85% | 191.53 | +19.16% |
| /wp-json/wc-analytics/leaderboards?after=2025-09-01T00%3A00%3A00&before=2025-09-02T23%3A59%3A59&per_page=5&persisted_query=%7B%7D&_locale=user | **164.26** | 164.84 | +0.35% | 176.52 | +7.46% |
| /wp-json/wc/v3/orders | **1229.57** | 1240.19 | +0.86% | 1263.75 | +2.78% |
| /wp-json/wc/v3/products | **277.26** | 282.47 | +1.88% | 317.49 | +14.51% |
| /wp-json/wc/v3/products/categories | 190.43 | **188.17** | **-1.19%** | 200.79 | +5.44% |
| /wp-json/wp/v2/block-patterns/patterns?_locale=user | **523.59** | 533.53 | +1.90% | 534.82 | +2.14% |
| /wp-json/wp/v2/blocks?context=edit&per_page=100&page=1&_locale=user | 221.50 | **207.84** | **-6.17%** | 212.51 | **-4.06%** |
| /wp-json/wp/v2/posts?_locale=user | 294.09 | 282.20 | **-4.04%** | **265.54** | **-9.71%** |
| /wp-json/wp/v2/settings?_locale=user | 218.27 | **212.02** | **-2.86%** | 221.06 | +1.28% |
| /wp-json/wp/v2/taxonomies?context=edit&per_page=100&_locale=user | 220.77 | 212.91 | **-3.56%** | **205.58** | **-6.88%** |
| /wp-json/wp/v2/templates?context=edit&per_page=100&_locale=user | 218.22 | **212.64** | **-2.56%** | 215.90 | **-1.06%** |
| /wp-json/wp/v2/templates/lookup?slug=front-page&_locale=user | 232.62 | **209.34** | **-10.01%** | 234.11 | +0.64% |
| /wp-json/wp/v2/users?context=view&who=authors&per_page=100&_fields=id%2Cname&_locale=user | 237.13 | 234.81 | **-0.98%** | **227.59** | **-4.02%** |
| /wp-json/wp/v2/users/1?context=view&_fields=id%2Cname&_locale=user | 231.08 | 230.50 | **-0.25%** | **213.94** | **-7.42%** |
| /wp-json/wp/v2/wp_pattern_category?context=view&per_page=100&_fields=id%2Cname%2Cdescription%2Cslug&_locale=user | 216.67 | 206.31 | **-4.78%** | **195.86** | **-9.60%** |

Overall, I'm not happy with current performance just doing the analytics.  The improvements aren't enough on some requests to necessarily warrant the degradation on others, especially with the analytics feature disabled.  

Here's what my current thoughts are for options to move forward:

1.  Migrate all the APIs at once, but as an experimental feature.
  - Keep the added `woocommerce_{$namespace}_rest_namespace_routes_registered` hooks so that other extensions can start to use those when they need to add to the existing namespaces in a way that is forward compatible.
  - Push for adoption over time.
    - Optionally remove some of the namespaces from behind the feature gate once we're more confident about compatibility.

2. Make a proposal to WordPress core to allow for lazy loading of namespaces.
  - Just the WP Core API namespaces take around 8-9ms to load in my tests regardless of the endpoint being hit.
  - API endpoints have a lot of overhead from translations that adds up as more endpoints are added.  
  - Should be able to register a namespace early and have custom hooks to load that namespace when needed.  Specific routes could always load all when needed for discovery.

3. Focus on fixing some of the reasons that currently make loading the `wc-analytics` routes slow, e.g. the use of `rest_do_request()` being run during route registration.

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

prettyboymp committed:
Wrap the inner logic of wc_rest_lazy_load_namespace to hide the $callback_filter_id parameter

prettyboymp committed:
fix lint issues

Konamiman commented:
> This PR introduces a new wc_rest_lazy_load_namespace() function

Could this be a class method instead of a function, please? We have that rule (not explicitly stated, but hinted in [the README for the includes directory](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/README.md)) that any new code should ideally go into classes in the `src` directory. Bare functions make the code very difficult to unit test.

prettyboymp committed:
Move the underlying function logic to the RestApiUtil class

prettyboymp committed:
switching back to $ns variables because $namespace is reserved

prettyboymp committed:
lint fixes

prettyboymp commented:
> > This PR introduces a new wc_rest_lazy_load_namespace() function
> 
> Could this be a class method instead of a function, please? We have that rule (not explicitly stated, but hinted in [the README for the includes directory](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/README.md)) that any new code should ideally go into classes in the `src` directory. Bare functions make the code very difficult to unit test.

@Konamiman, I moved the internal function to the `Utilities\RestApiUtil`, but left the main `wc_rest_lazy_load_namespace()` function.  My logic behind this is that I really hope that having WooCommerce handle the lazyloading is just temporary until WordPress core addresses https://core.trac.wordpress.org/ticket/63946 directly.

prettyboymp committed:
Fix container request for the RestApiUtil

Konamiman commented:
> My logic behind this is that I really hope that having WooCommerce handle the lazyloading is just temporary

An additional issue is that once we add a function we can't remove it and we need to keep its behavior for backwards compatibility. So if this code is for temporary use, it makes even more sense to have it only in classes in the iternal namespace.

prettyboymp commented:
> > My logic behind this is that I really hope that having WooCommerce handle the lazyloading is just temporary
> 
> An additional issue is that once we add a function we can't remove it and we need to keep its behavior for backwards compatibility. So if this code is for temporary use, it makes even more sense to have it only in classes in the iternal namespace.

If WordPress core adopted lazy loading namespaces, the function `wc_rest_lazy_load_namespace()` would still exist in WooCommerce Core, but be soft deprecated for a while.  It would just become a wrapper for WordPress core's lazy loading.

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

prettyboymp committed:
Updating namespace parameter to $route_namespace to match core's naming convention.

prettyboymp committed:
Drop the route parameter from the wrapping wc_rest_lazy_load_namespace() function.

prettyboymp committed:
lint fixes

prettyboymp committed:
Remove accidental change pending WP core work

Konamiman commented:
> Whether the function is a class method or global function doesn't change the way we would have to handle backward compatibility if this was moved to core.

By adding a new function we are making the functionality available to plugins and thus having to maintain compatibility. By adding it as a method in a class in the `Internal` namespace, however, we are free to alter, rename or even remove it in the future as we see fit. We can always take it out of `Internal` in the future if we decide that we _really_ want to make it available outside of WooCommerce core.

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

prettyboymp committed:
Making lazy loading completely internal for now.

prettyboymp committed:
adding strict types to test

prettyboymp commented on plugins/woocommerce/tests/php/src/Utilities/RestApiUtilTest.php:
```diff
@@ -0,0 +1,193 @@
+		// Should register filter but not execute callback immediately
+		$this->assertFalse( $callback_executed );
+
+		// Verify that rest_pre_dispatch filter was added
+		$this->assertTrue( has_filter( 'rest_pre_dispatch' ) );
+	}
```
`has_filter()` only returns boolean when no callback is given.

prettyboymp replied:
This is the documentation for `has_filter()`:

```php
/**
 * Checks if any filter has been registered for a hook.
 *
 * When using the `$callback` argument, this function may return a non-boolean value
 * that evaluates to false (e.g. 0), so use the `===` operator for testing the return value.
 *
 * @since 2.5.0
 *
 * @global WP_Hook[] $wp_filter Stores all of the filters and actions.
 *
 * @param string                      $hook_name The name of the filter hook.
 * @param callable|string|array|false $callback  Optional. The callback to check for.
 *                                               This function can be called unconditionally to speculatively check
 *                                               a callback that may or may not exist. Default false.
 * @return bool|int If `$callback` is omitted, returns boolean for whether the hook has
 *                  anything registered. When checking a specific function, the priority
 *                  of that hook is returned, or false if the function is not attached.
 */
```

prettyboymp committed:
lint fixes

prettyboymp committed:
more lint fixes

prettyboymp committed:
more ling

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

albarin commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,115 +55,167 @@ public function __construct() {
-				'Automattic\WooCommerce\Admin\API\Taxes',
-			);
+	/**
+	 * Load the wc-analytics namespace controllers.
```
Nit: should this say `wc-admin`?

mikejolley commented on plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/reports-coupons-stats.php:
```diff
@@ -36,6 +36,8 @@ public function setUp(): void {
 	 */
 	public function test_register_routes() {
+		// This namespace may be lazy loaded, so we make a discovery request to trigger loading for this test.
+		$this->server->dispatch( new WP_REST_Request( 'GET', '/' ) );
```
Instead of repeating this across all tests, could we add something to the abstract test class which does this and gets routes? It will make it less likely to be forgotten.

prettyboymp replied:
I don't think it's worth adding a new abstract class to the test hierarchy for this and it doesn't make sense adding this to the `WC_REST_Unit_Test_Case` class as it is lazy load namespace specific.  Adding it there would make it unclear as to whether that is a precedent to passing the tests for other classes.

mikejolley replied:
I don't know, the lazy loading seems like it should be something tests don't need to worry about.

albarin reviewed (approved):
✅ Confirmed the analytics bug is fixed.
✅ Smoke tested pages and posts with blocks like single product, category list, etc.
✅ Smoke tested the site editor, WC templates (shop, cart, checkout, product, etc.).
✅ Smoke WooCommerce Analytics within the admin.

mikejolley commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,115 +55,167 @@ public function __construct() {
+		}
+
+		$rest_api_util = wc_get_container()->get( RestApiUtil::class );
+		if ( is_callable( array( $rest_api_util, 'lazy_load_namespace' ) ) ) {
```
Under what circumstances would this not be callable?

prettyboymp replied:
👍 I was under the impression that our container allowed global instances to be replaced - which meant we couldn't immediately trust that an existing instance would have a new method.  But I was wrong.

Konamiman replied:
Side note: our DI container allows class replacements but only while running unit tests.

mikejolley commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,115 +55,167 @@ public function __construct() {
-				'Automattic\WooCommerce\Admin\API\MobileAppMagicLink',
-				'Automattic\WooCommerce\Admin\API\ShippingPartnerSuggestions',
-			);
+			$this->rest_api_init_wc_admin();
```
I'm not sure if I am misunderstanding the diff here, but wouldn't this need to use `lazy_load_namespace` in order to lazy load? `rest_api_init_wc_admin` registers the routes right away.

mikejolley replied:
If this is handled via `wc_rest_should_load_namespace` I think I'd want to understand when wc_rest_should_load_namespace/lazy_load_namespace is appropriate.

prettyboymp replied:
The goal on this diff changed and I didn't go back and update the content of the PR properly.  Initially I planned to lazy load both wc-admin and wc-analytics.  However, I was concerned about the number of extensions that may extend wc-admin compared to wc-analytics and decided to just start with wc-analytics for 10.3.

mikejolley commented on plugins/woocommerce/src/Utilities/RestApiUtil.php:
```diff
@@ -27,4 +27,89 @@ public function get_endpoint_data( $endpoint, $params = array() ) {
+	}
+
+	/**
+	 * This is the internal function that implements the logic of wc_rest_lazy_load_namespace(). Its interface
```
Where is `wc_rest_lazy_load_namespace`? This function does not exist. The PR description mentions it but its not implemented.

mikejolley commented on plugins/woocommerce/includes/rest-api/Server.php:
```diff
@@ -51,6 +51,12 @@ public function register_rest_routes() {
+			 *
+			 * @since 10.3.0
+			 */
+			do_action( 'woocommerce_' . $namespace . '_rest_namespace_routes_registered' );
```
Were these hooks introduced for a specific user need or reason? Just curious why they are needed.

prettyboymp replied:
@mikejolley, these were added because we're essentially changing the load order for these routes.  Any extension that is dependent on registering its routes after this namespace has registered will need to migrate to running on this hook rather than `rest_api_init`.

prettyboymp replied:
I'm going to remove this hook for now and revisit in a later PR.  If https://core.trac.wordpress.org/ticket/63946 gets adopted sooner than later, we can avoid adding this hook altogether.

mikejolley commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,115 +55,167 @@ public function __construct() {
 		if ( Features::is_enabled( 'launch-your-store' ) ) {
-			$controllers[] = 'Automattic\WooCommerce\Admin\API\LaunchYourStore';
+			$controller        = 'Automattic\WooCommerce\Admin\API\LaunchYourStore';
+			$this->$controller = new $controller();
```
What is going on here? Why are we assigning this controller to `$this->$controller`? 

Its not referenced anywhere else. Shouldn't we just call  `register_routes`? 

Which I might add is not lazy loaded either. Should it be?

prettyboymp replied:
> What is going on here? Why are we assigning this controller to $this->$controller?
> Its not referenced anywhere else. Shouldn't we just call register_routes?

This is just maintaining backward compatibility.  I have no idea why these were dynamically assigned to the class, but they were and they're automatically public, so there is risk in removing it.

> Which I might add is not lazy loaded either. Should it be?
No, I'm only lazy loading analytics for this PR.

mikejolley commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,115 +55,167 @@ public function __construct() {
+			 *
+			 * @since 3.5.0
+			 */
+			$controllers = apply_filters( 'woocommerce_admin_rest_controllers', $controllers );
```
Instead of conditionally calling a filter, but allowing all routes to be registered more than once, would it be make more sense to ensure `rest_api_init_wc_admin` is only called once? 

If its even needed. I wonder is this type of conditional filtering would cause race conditions with those consuming it.

prettyboymp replied:
The issue is that `rest_api_init_wc_admin` previously ran once after the wc-admin and wc-analytics controllers were both added to the $controllers array. So it was added to both `::rest_api_init_wc_admin()` and `::rest_api_init_wc_analytics()`. Running this filter twice, though, could cause extensions to insert their own controllers twice.  

Looking through the existing extensions that use this filter, they all just add to the $controllers array.  So this was a way to make sure these would still be added as needed for either namespace, but only once.

mikejolley replied:
Thats not difficult to mitigate if its just an array of controller names. TBH I think you should format this filter before iterating on it anyway, since it cannot be trusted. i.e. ensure its an array, ensure it has only strings, and ensure there are no duplicates.

mikejolley commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,115 +55,167 @@ public function __construct() {
+
+		foreach ( $controllers as $controller ) {
+			$this->$controller = new $controller();
+			$this->$controller->register_routes();
```
As above, why are we assigning these to class properties?

mikejolley commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -11,6 +11,7 @@
 defined( 'ABSPATH' ) || exit;
 
 use Automattic\WooCommerce\Internal\Admin\Loader;
```
`Loader` is unused.

mikejolley commented on plugins/woocommerce/src/Admin/API/Init.php:
```diff
@@ -54,115 +55,167 @@ public function __construct() {
+		 * @since 10.3.0
 		 */
-		$controllers = apply_filters( 'woocommerce_admin_rest_controllers', $controllers );
+		do_action( 'woocommerce_wc-admin_rest_namespace_routes_registered' ); // phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
```
Do we have to break naming conventions? We can sanitise `-` to `_` or pass as prop as an alternative.

prettyboymp replied:
Thanks for bringing this up.

I'm going to drop this filter for now and only keep the one for analytics since it is the only one we're initially lazy loading.   I'll go ahead and convert it to underscore.  I did bring this up in the upstream patch to WP Core: https://github.com/WordPress/wordpress-develop/pull/10080/files#r2398579660

mikejolley commented on plugins/woocommerce/src/Utilities/RestApiUtil.php:
```diff
@@ -27,4 +27,89 @@ public function get_endpoint_data( $endpoint, $params = array() ) {
+		}
+
+		if ( '' !== $rest_route ) {
+			$rest_route      = trailingslashit( ltrim( $rest_route, '/' ) );
```
Might be clearer to move this and the `$rest_route = $GLOBALS['wp']->query_vars['rest_route'] ?? '';` above to a private helper method.

mikejolley replied:
Still think this is useful. Coderabbit also suggested being more defensive with `$GLOBALS['wp']->query_vars` so you could do both.

mikejolley commented on plugins/woocommerce/src/Utilities/RestApiUtil.php:
```diff
@@ -27,4 +27,89 @@ public function get_endpoint_data( $endpoint, $params = array() ) {
+			};
+			$callback_filter_id = _wp_filter_build_unique_id( 'rest_pre_dispatch', $callback_filter, 0 );
+			// This runs on priority 0 so that the namespace is loaded before `rest_handle_options_request()` is run (priority 10).
+			add_filter( 'rest_pre_dispatch', $callback_filter, 0, 3 );
```
My understanding is that:

- We are passed a namespace and callback
- We create an anon function which calls `attach_lazy_loaded_namespace` with the namespace and callback
- `attach_lazy_loaded_namespace` calls the callback if the route matches the namespace, then unhooks itself

I *think* this is ok. My concern is the `recursive callback issue` mentioned in the comment above. 

Did you consider a registration type pattern whereby we hold on to the namespaces and callbacks and do a single filter on `rest_pre_dispatch` which iterates over registered callbacks and only calls those appropriate? This would mean recursion is not a problem at all.

prettyboymp replied:
> I think this is ok. My concern is the recursive callback issue mentioned in the comment above.

The recursion issue mentioned in the comment isn't from this function being recursive, but instead is because the wc-analytics namespace has recursive routes.  The route registration of some of the analytics controllers call `API\Reports\PerformanceIndicators::get_analytics_report_data()` when building the parameter structure.  This method ends up creating a rest request to itself to get the data it needs... which then made multiple requests to itself because of the self ref links needed to be generated (see https://github.com/woocommerce/woocommerce/pull/60791).

mikejolley replied:
Could you add some explanation inline? This is the type of thing we'll come back to in a few months and be stumped by.

mikejolley reviewed (changes requested):
I've left some feedback. Its possible some things were changed after the initial PR making the description outdated.

prettyboymp committed:
Fix documentation for method

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

prettyboymp committed:
Addressing feedback:
- dropped is_callable check against the RestApiUtil
- dropped the woocommerce_wc-admin_rest_namespace_routes_registered filter as it isn't currently being lazy loaded.
- converted analytics filter to woocommerce_wc_analytics_rest_namespace_routes_registered to get rid of underscore convention change.

prettyboymp committed:
Drop the woocommerce_' . $namespace . '_rest_namespace_routes_registered hook from the general WooCommerce Rest API server

prettyboymp committed:
removing unneeded use

prettyboymp commented:
> I've left some feedback. Its possible some things were changed after the initial PR making the description outdated.

@mikejolley Thank you for reviewing. I updated the description a bit as things did change as I worked on the PR more.  There is no longer a global function for `wc_rest_lazy_load_namespace()` and we're leaving it internal for now.  That limits the impact for having other extensions use lazy loading, but hopefully an upstream option will land soon enough.  Ideally, though, any newly created REST namespaces can use the new method in the interim to avoid further adding a performance hit to all API requests.

prettyboymp committed:
Dropping all the action hooks for now.
- There isn't an immediate need for them.

mikejolley reviewed (approved):
Thanks for taking care of the main feedback. I'm going to pre-approve so my previous review is not blocking this. I do still think there is value in being more defensive particularly around filtered values.

prettyboymp committed:
Added hardening around controller filter return values.
Removed the duplicate run check against the woocommerce_admin_rest_controllers and opted for a second filter for analyitics. Avoids breaking backward compatibility since admin won't be switched to lazy loading any time zoon.

prettyboymp committed:
Update comment.

prettyboymp committed:
hardening around usage of global WP instance.

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-2074/load-wc-admin-api-on-demand

prettyboymp committed:
Switch back to using a shared hook between analyitcs and wc-admin controllers.
